### PR TITLE
[core] Recover spilled objects that are lost during node failure

### LINKED ISSUE #17377: [lineage] Do not reconstruct objects that are spilled
### Describe your feature request

Objects that are spilled with no in-memory locations are currently considered lost. We should instead try to reload them into memory on demand.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Currently there is a bug that causes failure handling to not trigger correctly when a spilled object is lost. This PR modifies object recovery so that it triggers from the ReferenceCounter, whenever the primary or spilled location of an object is lost. If this happens, then the object ID is added to a buffer that is periodically flushed and recovered by the main CoreWorker thread.

## Related issue number

Closes #17377.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
TODO: Add C++ unit tests, but the rest of it is ready for review.

File: src/ray/core_worker/reference_count.h [hunk 1]
```diff
@@ -434,7 +440,7 @@ class ReferenceCounter : public ReferenceCounterInterface,
   /// \param[in] release Whether to release the reference.
```
Comment (scv119) [inline src/ray/core_worker/reference_count.h:440]: this comment no longer needed?

File: src/ray/core_worker/reference_count.cc [hunk 2]
```diff
@@ -1167,29 +1188,36 @@ size_t ReferenceCounter::GetObjectSize(const ObjectID &object_id) const {
...
-    it->second.spilled_node_id = spilled_node_id;
-  }
-  if (size > 0) {
-    it->second.object_size = size;
-  }
-  PushToLocationSubscribers(it);
-  if (release) {
-    // Release the primary plasma copy, if any.
+  bool spilled_location_alive =
+      spilled_node_id.IsNil() || check_node_alive_(spilled_node_id);
```
Comment (scv119) [inline src/ray/core_worker/reference_count.cc:1204]: does spilled_node_id IsNil mean it's spilled to cloud?

**Review by scv119 — COMMENTED** [review]
the reference counter part looks good to me

File: src/ray/core_worker/reference_count.cc [hunk 3]
```diff
@@ -1167,29 +1188,36 @@ size_t ReferenceCounter::GetObjectSize(const ObjectID &object_id) const {
...
-                                           const NodeID &spilled_node_id, int64_t size,
-                                           bool release) {
+                                           const NodeID &spilled_node_id, int64_t size) {
   absl::MutexLock lock(&mutex_);
   auto it = object_id_refs_.find(object_id);
   if (it == object_id_refs_.end()) {
     RAY_LOG(WARNING) << "Spilled object " << object_id << " already out of scope";
     return false;
   }
+  if (it->second.OutOfScope(lineage_pinning_enabled_)) {
```
Comment (rkooo567) [inline src/ray/core_worker/reference_count.cc:1198]: This seems to be unrelated to this PR. Is it a bug you discovered or sth?

File: src/ray/core_worker/reference_count.cc [hunk 4]
```diff
@@ -628,18 +637,25 @@ bool ReferenceCounter::SetDeleteCallback(
...
-    const NodeID &raylet_id) {
+void ReferenceCounter::ResetObjectsOnRemovedNode(const NodeID &raylet_id) {
   absl::MutexLock lock(&mutex_);
-  std::vector<ObjectID> lost_objects;
   for (auto it = object_id_refs_.begin(); it != object_id_refs_.end(); it++) {
     const auto &object_id = it->first;
-    if (it->second.pinned_at_raylet_id.value_or(NodeID::Nil()) == raylet_id) {
-      lost_objects.push_back(object_id);
+    if (it->second.pinned_at_raylet_id.value_or(NodeID::Nil()) == raylet_id ||
+        it->second.spilled_node_id == raylet_id) {
```
Comment (rkooo567) [inline src/ray/core_worker/reference_count.cc:645]: Isn't `it->second.spilled_node_id` always equivalent to `it->second.pinned_at_raylet_id`? Is this if predicate necessary? (If they are not the same, it could be a bug because we release a pinned object when the plasma object is spilled?)

File: src/ray/core_worker/reference_count.cc [hunk 5]
```diff
@@ -599,6 +599,15 @@ void ReferenceCounter::ReleasePlasmaObject(ReferenceTable::iterator it) {
     it->second.on_delete = nullptr;
   }
   it->second.pinned_at_raylet_id.reset();
+  if (it->second.spilled) {
+    // The spilled copy of the object should get deleted during the on_delete
+    // callback, so reset the spill location metadata here.
+    // TODO(swang): Spilled copies in cloud storage are not GCed, so we should
```
Comment (rkooo567) [inline src/ray/core_worker/reference_count.cc:605]: Maybe we should just add logic now? Isn't it just spilled & node_id == Nil, do nothing?

**Review by rkooo567 — COMMENTED** [review]
LGTM! Just a few comments. is there a particular reason why we change from event-based logic (when node fails) -> internal loop to recover objects?

File: src/ray/core_worker/reference_count.cc (hunk 2 above)
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.cc:1204]: Yep!

File: src/ray/core_worker/reference_count.cc (hunk 3 above)
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.cc:1198]: Actually it is related. We shouldn't add the spilled location if the object is already out of scope because it means the spilled copy is already being deleted by the raylet. I'll add a note to explain this.

File: src/ray/core_worker/reference_count.cc (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.cc:645]: Hmm I think you're right but I think this is a bit easier to understand.

**Comment by stephanie-wang** [comment]
> LGTM! Just a few comments. is there a particular reason why we change from event-based logic (when node fails) -> internal loop to recover objects?

Yeah I did this to prevent weird callback relationships between the ref counter and the recovery manager. The recovery manager already calls the ref counter, so instead of adding a reverse callback to the ref counter to recover lost spilled objects, I used the event loop style.

**Comment by stephanie-wang** [comment]
Also, this PR is currently blocked on #21719 so I'm waiting on that one first.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by rkooo567** [comment]
Is it ready to review again? Seems like there are many failing tests for some reasons!

**Comment by stephanie-wang** [comment]
> Is it ready to review again? Seems like there are many failing tests for some reasons!

Yes, it's ready for review. Working on the tests.

**Review by rkooo567 — APPROVED** [review]

File: src/ray/core_worker/reference_count.cc (hunk 4 above)
Comment (rkooo567) [reply src/ray/core_worker/reference_count.cc:636]: Maybe I prefer `RAY_CHECK(it->second.spilled_node_id == raylet_id)` (in case this invariant changes), but it is not a strong requirements.

**Review by scv119 — APPROVED** [review]
